### PR TITLE
Add support for generic http client handler.

### PR DESCRIPTION
**Pull request by rghetia**

### CODE REVIEW

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 1]
```diff
@@ -103,4 +103,31 @@ public void handleEnd(Span span, @Nullable P response, @Nullable Throwable error
     span.setStatus(HttpTraceUtil.parseResponseStatus(statusCode, error));
     span.end();
   }
+
+  String getSpanName(Q request, HttpExtractor<Q, P> extractor) {
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: final?

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 2]
```diff
@@ -103,4 +103,31 @@ public void handleEnd(Span span, @Nullable P response, @Nullable Throwable error
...
+  }
+
+  private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullable String value) {
+    if (value != null && !value.isEmpty()) {
+      span.putAttribute(key, AttributeValue.stringAttributeValue(value));
+    }
+  }
+
+  void addSpanRequestAttributes(Span span, Q request, HttpExtractor<Q, P> extractor) {
+    putAttributeIfNotEmptyOrNull(span, "http.user_agent", extractor.getUserAgent(request));
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: Strings like "http.user_agent" please make them constants

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 3]
```diff
@@ -103,4 +103,31 @@ public void handleEnd(Span span, @Nullable P response, @Nullable Throwable error
...
+    return path;
+  }
+
+  private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullable String value) {
+    if (value != null && !value.isEmpty()) {
+      span.putAttribute(key, AttributeValue.stringAttributeValue(value));
+    }
+  }
+
+  void addSpanRequestAttributes(Span span, Q request, HttpExtractor<Q, P> extractor) {
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: final?

File: contrib/http_util/src/test/java/io/opencensus/contrib/http/AbstractHttpHandlerTest.java [hunk 4]
```diff
@@ -105,4 +120,33 @@ public void handleEndShouldEndSpan() {
...
+
+  private void verifyAttributes(String key) {
+    verify(span).putAttribute(eq(key), attributeCaptor.capture());
+    AttributeValue attribute = attributeCaptor.getValue();
+    assertThat(attribute.toString()).contains(attributeMap.get(key));
+  }
+
+  @Test
+  public void testSpanRequestAttributes() {
+
```
Comment (bogdandrutu) [inline contrib/http_util/src/test/java/io/opencensus/contrib/http/AbstractHttpHandlerTest.java]: Remove empty line.

File: contrib/http_util/src/test/java/io/opencensus/contrib/http/util/testing/FakeSpan.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import io.opencensus.trace.Link;
+import io.opencensus.trace.MessageEvent;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.Status;
+import java.util.Map;
+
+/** A fake {@link Span} which allows user to set {@link SpanContext} upon initialization. */
+@Internal
+public class FakeSpan extends Span {
```
Comment (bogdandrutu) [inline contrib/http_util/src/test/java/io/opencensus/contrib/http/util/testing/FakeSpan.java]: Why is this needed?

Comment (rghetia) [reply contrib/http_util/src/test/java/io/opencensus/contrib/http/util/testing/FakeSpan.java]: Testing is more simpler with FakeSpan then mocking. In order to test tracecontext injection you would need to mock Span, SpanContext and several methods. I see that we have used TestSpan for log correlation.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpClientHandler.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    checkNotNull(carrier, "carrier");
+    checkNotNull(request, "request");
+    if (parent == null) {
+      parent = tracer.getCurrentSpan();
+    }
+    String spanName = getSpanName(request, extractor);
+    SpanBuilder builder = tracer.spanBuilderWithExplicitParent(spanName, parent);
+    Span span = builder.startSpan();
+
+    addSpanRequestAttributes(span, request, extractor);
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpClientHandler.java]: Protect all the calls to addAttriputes/Annotations/MessageEvents with Span.isRecordEvents.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 1 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: done.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 3 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: done.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 2 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: done.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpClientHandler.java (hunk 6 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpClientHandler.java]: done.

File: contrib/http_util/src/test/java/io/opencensus/contrib/http/AbstractHttpHandlerTest.java (hunk 4 above)
Comment (rghetia) [reply contrib/http_util/src/test/java/io/opencensus/contrib/http/AbstractHttpHandlerTest.java]: done

**Comment by rghetia** [comment]
@bogdandrutu PTAL.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 7]
```diff
@@ -24,10 +24,18 @@
 import io.opencensus.trace.MessageEvent;
 import io.opencensus.trace.MessageEvent.Type;
 import io.opencensus.trace.Span;
+import io.opencensus.trace.Span.Options;
 import javax.annotation.Nullable;
 
 /** Base class for handling request on http client and server. */
 abstract class AbstractHttpHandler<Q, P> {
+  static final String HTTP_HOST = "http.host";
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: Consider to have all these in a public file (constants) if people want to share the same values.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 8]
```diff
@@ -99,8 +111,37 @@ public final void handleMessageReceived(Span span, long messageId, long messageS
...
+    return path;
+  }
+
+  private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullable String value) {
+    if (value != null && !value.isEmpty()) {
+      span.putAttribute(key, AttributeValue.stringAttributeValue(value));
+    }
+  }
+
+  final void addSpanRequestAttributes(Span span, Q request, HttpExtractor<Q, P> extractor) {
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: if (span.getOptions().contains(Options.RECORD_EVENTS))  { ... } ? You will call this on the server as well.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 7 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: sure.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 8 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: yes.
